avalon/tools/widgets.py

```diff
@@ -322,3 +322,167 @@ def _list_project_silos():
+        widget = OptionalActionWidget(self.label, parent)
+
+        if self.icon:
+            widget.setIcon(self.icon)
+
+        if self.use_option:
+            widget.option.clicked.connect(self.on_option)
+            widget.option.setToolTip(self.option_tip)
+        else:
+            widget.option.setVisible(False)
+
+        return widget
+
+    def set_option_tip(self, options):
+        sep = "\n\n"
+        mak = (lambda opt: opt["name"] + " :\n    " + opt["help"])
+        self.option_tip = sep.join(mak(opt) for opt in options)
+
+    def on_option(self):
+        self.optioned = True
+
+
+class OptionalActionWidget(QtWidgets.QWidget):
+    """Main widget class for `OptionalAction`"""
+
+    def __init__(self, label, parent=None):
+        super(OptionalActionWidget, self).__init__(parent)
+
+        body = QtWidgets.QWidget()
+        body.setStyleSheet("background: transparent;")
+
+        icon = QtWidgets.QLabel()
+        label = QtWidgets.QLabel(label)
+        option = OptionBox(body)
+
+        icon.setFixedSize(24, 16)
+        option.setFixedSize(30, 30)
+
+        layout = QtWidgets.QHBoxLayout(body)
+        layout.setContentsMargins(0, 0, 0, 0)
+        layout.setSpacing(2)
+        layout.addWidget(icon)
+        layout.addWidget(label)
+        layout.addSpacing(6)
+
+        layout = QtWidgets.QHBoxLayout(self)
+        layout.setContentsMargins(6, 1, 2, 1)
+        layout.setSpacing(0)
+        layout.addWidget(body)
+        layout.addWidget(option)
+
+        self.setFixedHeight(32)
+        self.setMouseTracking(True)
+
+        self.icon = icon
+        self.option = option
+        self.body = body
+
+        # (NOTE) For removing ugly QLable shadow FX when highlighted in Nuke.
+        #   See https://stackoverflow.com/q/52838690/4145300
+        label.setStyle(QtWidgets.QStyleFactory.create("Plastique"))
+
+    def setIcon(self, icon):
+        pixmap = icon.pixmap(16, 16)
+        self.icon.setPixmap(pixmap)
+
+    def enterEvent(self, event):
+        self.body.setBackgroundRole(QtGui.QPalette.Highlight)
+        self.body.setAutoFillBackground(True)
+
+    def leaveEvent(self, event):
+        self.body.setBackgroundRole(QtGui.QPalette.Window)
+        self.body.setAutoFillBackground(False)
+
+
+class OptionBox(QtWidgets.QWidget):
+    """Option box widget class for `OptionalActionWidget`"""
+
+    clicked = QtCore.Signal()
+
+    def __init__(self, parent):
+        super(OptionBox, self).__init__(parent)
+
+        label = QtWidgets.QLabel()
+
+        layout = QtWidgets.QHBoxLayout(self)
+        layout.setContentsMargins(0, 0, 0, 0)
+        layout.addSpacing(8)
+        layout.addWidget(label)
+
+        icon = qtawesome.icon("fa.sticky-note-o", color="#c6c6c6")
+        pixmap = icon.pixmap(18, 18)
+        label.setPixmap(pixmap)
+
+        self.setStyleSheet("background: transparent;")
+        self._hovered = False
+
+    def mouseReleaseEvent(self, event):
+        if self._hovered:
```

Comment:
Small detail, enter and leave events are ignored when mouse is pressed (as far as I know) so I would change this condition a little bit:
```
mouse_pos = event.pos()
if self._hovered and self.rect().contains(mouse_pos):
    ...
```

Reply:
Thanks @iLLiCiTiT , I had a quick test on this but not able to see the difference, maybe it’s related to how I implemented option box with `QWidgetAction`, not sure. I’ll keep this in mind and submit another PR if needed.

Reply:
It's just small detail :) for cases when you press mouse over the button but then you change your mind. With this modification is possible to leave from button with pressed mouse and release to not trigger the `clicked` signal.

Reply:
Although not directly related to this, but after 964de13 and tested, I am pretty sure that one can have second thoughts after mouse pressed !